runtime/executor/src/main/java/org/apache/nemo/runtime/executor/datatransfer/BlockOutputWriter.java

```diff
@@ -100,12 +139,77 @@ public void writeWatermark(final Watermark watermark) {
   }
 
   /**
-   * Notifies that all writes for a block is end.
-   * Further write about a committed block will throw an exception.
+   * Notifies that all writes for a block have ended.
+   * Further write to a committed block will throw an exception.
    */
   @Override
   public void close() {
+    // for MemoryStore blocks
+    if (blockStoreValue == DataStoreProperty.Value.MEMORY_STORE
+      || blockStoreValue == DataStoreProperty.Value.SERIALIZED_MEMORY_STORE) {
+      LOG.info("BOW close called, iterating over STV and writing to block");
+      if (memoryManager.acquireStorageMemory(blockToWrite.getId(), sizeTrackingVector.estimateSize())) {
+        sizeTrackingVector.forEach(element -> blockToWrite.write(partitioner.partition(element), element));
+//        for (Object element : sizeTrackingVector) {
+//          blockToWrite.write(partitioner.partition(element), element);
+//        }
+      } else { // block does not fit in memory
+        LOG.info("Not enough storage memory, remaing StoragePool Memory: {}, size of block is {}, blockID: {}",
+          memoryManager.getRemainingStorageMemory(), sizeTrackingVector.estimateSize(), blockToWrite.getId());
+//        blockStoreValue = DataStoreProperty.Value.LOCAL_FILE_STORE;
+//        create new file block
+//        String newBlockId = blockToWrite.getId();
+//        BlockStore previousBlockStore = blockManagerWorker.getBlockStore(DataStoreProperty.Value.LOCAL_FILE_STORE);
+//        LOG.info("try to delete old block");
+//        previousBlockStore.deleteBlock(blockToWrite.getId());
+//        BlockStore newBlockStore = blockManagerWorker.getBlockStore(DataStoreProperty.Value.LOCAL_FILE_STORE);
+//        Block newBlock = newBlockStore.createBlock(newBlockId);
+//        LOG.info("writing to new block, newblockid {}", newBlock.getId());
+//        sizeTrackingVector.forEach(element -> newBlock.write(partitioner.partition(element), element));
+//        blockToWrite = newBlock;
+//        runtimeEdge.changeDataStoreProperty(DataStoreProperty.Value.LOCAL_FILE_STORE);
+//        BlockStore fileBLockStore = blockManagerWorker.getBlockStore(DataStoreProperty.Value.LOCAL_FILE_STORE);
+        /// dongjoo: change id + spilled
+//        Block newBlock = blockManagerWorker.createBlock(blockToWrite.getId(),
+//          DataStoreProperty.Value.LOCAL_FILE_STORE);
+//        blockStoreValue = DataStoreProperty.Value.LOCAL_FILE_STORE;
+//        block
+//        sizeTrackingVector.forEach(element -> blockToWrite.write(partitioner.partition(element), element));
+        // populate newly created block
+        sizeTrackingVector.forEach(element ->
+          potentialSpilledBlocktoWrite.write(partitioner.partition(element), element));
+//        sizeTrackingVector.forEach(element -> newBlock.write(partitioner.partition(element), element));
+        blockManagerWorker.putSpilledBlock(blockToWrite, potentialSpilledBlocktoWrite);
```

Comment:
Let's discuss about this way of spilling also - I think spilling Block(create a bogus potentialBlock that isn't in PhysicalPlan, store data in it, and replace original Block with this potentialBlock) will definitely introduce complex consistency issues.